app/services/ProcessCsvService.scala

```diff
@@ -0,0 +1,186 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.HttpResponse
+import akka.stream.alpakka.csv.scaladsl.CsvParsing
+import akka.stream.scaladsl.{Flow, Sink, Source}
+import akka.util.ByteString
+import config.ApplicationConfig
+import controllers.auth.RequestWithOptionalEmpRef
+import models.upscan.{UploadedSuccessfully, UpscanCsvFilesCallback, UpscanCsvFilesCallbackList}
+import models.{ERSFileProcessingException, RowValidationResults, SheetErrors}
+import org.apache.commons.io.FilenameUtils
+import play.api.Logger
+import play.api.i18n.Messages
+import play.api.mvc.AnyContent
+import services.FlowOps.eitherFromFunction
+import services.validation.ErsValidator.getCells
+import uk.gov.hmrc.services.validation.models._
+import uk.gov.hmrc.http.{HeaderCarrier, UpstreamErrorResponse}
+import uk.gov.hmrc.services.validation.DataValidator
+import utils.{CsvParserUtil, ERSUtil}
+
+import javax.inject.{Inject, Singleton}
+import scala.annotation.tailrec
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+@Singleton
+class ProcessCsvService @Inject()(parserUtil: CsvParserUtil,
+                                  dataGenerator: DataGenerator,
+                                  appConfig: ApplicationConfig,
+                                  ersUtil: ERSUtil
+                                 )(implicit executionContext: ExecutionContext,
+                                   actorSystem: ActorSystem) {
+
+  private val uploadCsvSizeLimit: Int = appConfig.uploadCsvSizeLimit
+
+  def extractEntityData(response: HttpResponse): Source[ByteString, _] =
+    response match {
+      case HttpResponse(akka.http.scaladsl.model.StatusCodes.OK, _, entity, _) => entity.withSizeLimit(uploadCsvSizeLimit).dataBytes
+      case notOkResponse =>
+        Source.failed(
+          UpstreamErrorResponse(
+            s"[ProcessCsvService][extractEntityData] Illegal response from Upscan: ${notOkResponse.status.intValue}, body: ${notOkResponse.entity.dataBytes}",
+            notOkResponse.status.intValue))
+    }
+
+  def extractBodyOfRequest: Source[HttpResponse, _] => Source[Either[Throwable, List[ByteString]], _] =
+    _.flatMapConcat(extractEntityData)
+      .via(CsvParsing.lineScanner())
+      .via(Flow.fromFunction(Right(_)))
+      .recover {
+        case e => Left(e)
+      }
+
+  def processFiles(callback: Option[UpscanCsvFilesCallbackList], scheme: String, source: String => Source[HttpResponse, _])(
+    implicit request: RequestWithOptionalEmpRef[AnyContent], hc: HeaderCarrier, messages: Messages
+  ): List[Future[Either[Throwable, Boolean]]] =
+    callback.get.files map { file =>
```

Comment:
Is this .get safe? I would assume if an optional value is being fed in here, there should be some logic for a None case, but there isn't one.

Reply:
it's not safe, but we didn't have the time to change this - this is upscan-related functionality rather than strictly processing; this part of the functionality follows on from here: https://github.com/hmrc/ers-checking-frontend/blob/master/app/controllers/UploadController.scala#L110, it's old functionality in a new space